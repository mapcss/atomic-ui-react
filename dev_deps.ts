export {
  anyBoolean,
  anyFunction,
  anyNumber,
  anyString,
  defineExpect,
  defineGlobalThis,
  fn,
  jestMatcherMap,
  jestModifierMap,
} from "https://deno.land/x/unitest@v1.0.0-beta.82/mod.ts";
import {
  defineExpect,
  equal,
  jestMatcherMap,
  jestModifierMap,
  MatchResult,
} from "https://deno.land/x/unitest@v1.0.0-beta.82/mod.ts";
export { FakeTime } from "https://deno.land/std@0.136.0/testing/time.ts";
export { describe, it } from "https://deno.land/std@0.136.0/testing/bdd.ts";
export { assertSnapshot } from "https://deno.land/std@0.136.0/testing/snapshot.ts";
export { default as clsx } from "https://esm.sh/clsx@1.1.1?pin=v78";
export async function setupJSDOM(): Promise<void> {
  const { JSDOM } = await import("https://esm.sh/jsdom@19.0.0?pin=v76").then((
    module,
  ) => module.default);

  const doc = new JSDOM(`<!DOCTYPE html>`);
  globalThis.document = doc.window.document;
  globalThis.HTMLIFrameElement = doc.window.HTMLIFrameElement;
  globalThis.Node = doc.window.Node;
}
export const expect = defineExpect({
  matcherMap: {
    ...jestMatcherMap,
    toHaveFocus,
    toHaveAttribute,
  },
  modifierMap: jestModifierMap,
});

function toHaveFocus(value: Element): MatchResult {
  return {
    pass: value.ownerDocument.activeElement === value,
    expected: "Actual element is focused",
  };
}

function toHaveAttribute(
  el: Element,
  name: string,
  expected: unknown,
): MatchResult {
  if (!el.hasAttribute(name)) {
    return {
      pass: false,
      expected: `[${name}] is not exists`,
    };
  }
  return {
    pass: equal(el.getAttribute(name), expected),
    expected: `${name} is ${Deno.inspect(expected)}`,
  };
}

export function setupRaf(): () => void {
  Object.defineProperty(globalThis.window, "requestAnimationFrame", {
    value: (handler: TimerHandler) => globalThis.setTimeout(handler, 1),
    configurable: true,
    writable: true,
  });

  Object.defineProperty(globalThis.window, "cancelAnimationFrame", {
    value: globalThis.clearTimeout,
    configurable: true,
    writable: true,
  });

  const reset = (): void => {
    globalThis.requestAnimationFrame = _.requestAnimationFrame;
    globalThis.cancelAnimationFrame = _.cancelAnimationFrame;
  };

  return reset;
}

// deno-lint-ignore no-explicit-any
export type ParamReturn<T extends (...args: any[]) => unknown> = [
  ...Parameters<T>,
  ReturnType<T>,
];

const _: {
  getComputedStyle: typeof window.getComputedStyle;
  requestAnimationFrame: typeof globalThis.requestAnimationFrame;
  cancelAnimationFrame: typeof cancelAnimationFrame;
} = {
  getComputedStyle: globalThis.getComputedStyle,
  requestAnimationFrame: globalThis.requestAnimationFrame,
  cancelAnimationFrame: globalThis.cancelAnimationFrame,
};

export function mockGetComputedStyle(mock: () => CSSStyleDeclaration) {
  Object.defineProperty(globalThis.window, "getComputedStyle", {
    value: mock,
    writable: true,
    configurable: true,
  });

  const reset = (): void => {
    globalThis.getComputedStyle = _.getComputedStyle;
  };

  return reset;
}
