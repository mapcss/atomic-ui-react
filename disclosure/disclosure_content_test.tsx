import { assertSnapshot, describe, it, setupJSDOM } from "../dev_deps.ts";
import { ReactElement } from "react";
import DisclosureContent from "./disclosure_content.ts";
import DisclosureProvider from "./disclosure_provider.ts";
import SSRProvider from "../ssr/ssr_provider.ts";
import { render } from "@testing-library/react";

const describeTests = describe({
  name: "DisclosureContent",
  async beforeEach() {
    await setupJSDOM();
  },
});

it(describeTests, "should be render as", (t) => {
  const { container, rerender } = render(
    <DisclosureContent>test</DisclosureContent>,
    {
      wrapper: ({ children }) => (
        <SSRProvider>
          <DisclosureProvider>{children as ReactElement}</DisclosureProvider>
        </SSRProvider>
      ),
    },
  );

  assertSnapshot(t, container.innerHTML);

  rerender(<DisclosureContent as="p">test</DisclosureContent>);
  assertSnapshot(t, container.innerHTML);
  rerender(<DisclosureContent className="test">test</DisclosureContent>);
  assertSnapshot(t, container.innerHTML);

  rerender(
    <DisclosureContent style={{ display: "unset" }}>test</DisclosureContent>,
  );
  assertSnapshot(t, container.innerHTML);

  rerender(
    <DisclosureContent
      className="test"
      renderAttributes={() => ({
        className: "test2",
      })}
    >
      test
    </DisclosureContent>,
  );
  assertSnapshot(t, container.innerHTML);

  rerender(
    <DisclosureContent
      renderAttributes={() => ({
        style: {
          display: undefined,
        },
      })}
    >
      test
    </DisclosureContent>,
  );
  rerender(
    <DisclosureContent
      renderAttributes={({ isOpen }) => ({
        className: isOpen ? "open" : "close",
      })}
    >
      test
    </DisclosureContent>,
  );
  assertSnapshot(t, container.innerHTML);
});

it(describeTests, "should be render if disclosure open", (t) => {
  const { container } = render(
    <DisclosureContent>test</DisclosureContent>,
    {
      wrapper: ({ children }) => (
        <SSRProvider>
          <DisclosureProvider isDefaultOpen>
            {children as ReactElement}
          </DisclosureProvider>
        </SSRProvider>
      ),
    },
  );

  assertSnapshot(t, container.innerHTML);
});
