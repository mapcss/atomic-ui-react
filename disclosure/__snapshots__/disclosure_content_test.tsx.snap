export const snapshot = {};

snapshot[`DisclosureContent > should be render as 1`] = `'<div id="atomic-ui-0" hidden="">test</div>'`;

snapshot[`DisclosureContent > should be render as 2`] = `'<p id="atomic-ui-0" hidden="">test</p>'`;

snapshot[`DisclosureContent > should be render as 3`] = `'<div id="atomic-ui-0" hidden="" class="test">test</div>'`;

snapshot[`DisclosureContent > should be render as 4`] = `'<div id="atomic-ui-0" hidden="" style="display: unset;">test</div>'`;

snapshot[`DisclosureContent > should be render as 5`] = `'<div id="atomic-ui-0" hidden="" style="" class="test test2">test</div>'`;

snapshot[`DisclosureContent > should be render as 6`] = `'<div id="atomic-ui-0" hidden="" style="" class="close">test</div>'`;

snapshot[`DisclosureContent > should be render if disclosure open 1`] = `'<div id="atomic-ui-0">test</div>'`;
